### PR TITLE
Add WebuiHelper#user_and_role tests

### PR DESCRIPTION
**Pull request by mschnitzer**

### CODE REVIEW

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 1]
```diff
@@ -319,7 +319,69 @@
...
+    context 'with options' do
+      it 'does not show an icon if option disables it' do
+        expect(user_and_role(user.login, nil, no_icon: true)).to eq(
+          "<a href=\"/user/show/#{user.login}\">#{user.realname} (#{user.login})</a>"
+        )
+      end
+
+      it 'only shows user login if short option is set' do
+        user.realname = ''
+        user.save!
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Why is this necessary?

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 2]
```diff
@@ -319,7 +319,69 @@
...
   describe '#user_and_role' do
-    skip('Please add some tests')
+    let(:user) { create(:user) }
+    let(:nobody) { create(:user_nobody) }
+
+    before do
+      User.current = user
+    end
+
+    context 'for users' do
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: The whole test is for users, right? So this should be sth like ``for logged in users``?

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 3]
```diff
@@ -319,7 +319,69 @@
...
+        User.current = nobody
+
+        expect(user_and_role(user.login)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user.realname}\" " + \
+          "src=\"/user/icon/#{user.login}?size=20\" />#{user.realname} (#{user.login})"
+        )
+      end
+    end
+
+    context 'with roles' do
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: This context basically belongs to the ``for users`` context

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 4]
```diff
@@ -319,7 +319,69 @@
...
+        expect(user_and_role(user.login, nil, short: true)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user}\" src=\"/user/icon/#{user.login}?size=20\" " + \
+          "/><a href=\"/user/show/#{user.login}\">#{user.login}</a>"
+        )
+      end
+    end
+
+    context 'for guests' do
+      it 'does not link to user profiles if current user is nobody' do
+        User.current = nobody
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: This can be done in a before filter

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 5]
```diff
@@ -319,7 +319,69 @@
   end
 
   describe '#user_and_role' do
-    skip('Please add some tests')
+    let(:user) { create(:user) }
+    let(:nobody) { create(:user_nobody) }
+
+    before do
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: This belongs inside the ``for users`` context

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 3 above)
Comment (mschnitzer) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Why? It's not necessary to be logged in to see the roles.

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 1 above)
Comment (mschnitzer) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: You would get
```
       expected: "<img width=\"20\" height=\"20\" alt=\"user_1\" src=\"/user/icon/user_1?size=20\" />user_1"
            got: "<img width=\"20\" height=\"20\" alt=\"Seamus Baumbach Sr.\" src=\"/user/icon/user_1?size=20\" />user_1"
```
otherwise.

Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: and why don't you use `user.realname` instead of `user` in your expect string? :confused:  Then you don't need to set it to empty.

Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: and why don't you use `user.realname` instead of `user` in your expect string? :confused:  Then you don't need to set it to empty.

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 6]
```diff
@@ -319,7 +319,79 @@
   end
 
   describe '#user_and_role' do
-    skip('Please add some tests')
+    let(:user) { create(:user) }
+    let(:nobody) { create(:user_nobody) }
```
Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: why do yo need this?

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 7]
```diff
@@ -319,7 +319,79 @@
...
+        expect(user_and_role(user.login)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user}\" src=\"/user/icon/#{user.login}?size=20\" " + \
+          "/><a href=\"/user/show/#{user.login}\">#{user.login}</a>"
+        )
+      end
+    end
+
+    context 'with options' do
+      before do
+        User.current = nobody
```
Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: why don't set it to `nil`?

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 8]
```diff
@@ -319,7 +319,79 @@
...
+        expect(user_and_role(user.login, nil, short: true)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user}\" src=\"/user/icon/#{user.login}?size=20\" " + \
+          "/>#{user.login}"
+        )
+      end
+    end
+
+    context 'for guests' do
+      before do
+        User.current = nobody
```
Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: same here!

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 9]
```diff
@@ -319,7 +319,79 @@
...
+        expect(user_and_role(user.login)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user.realname}\" " + \
+          "src=\"/user/icon/#{user.login}?size=20\" />#{user.realname} (#{user.login})"
+        )
+      end
+    end
+
+    context 'with roles' do
+      before do
+        User.current = nobody
```
Comment (Ana06) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: and here

File: src/api/spec/helpers/webui/webui_helper_spec.rb [hunk 10]
```diff
@@ -319,7 +319,76 @@
...
+        expect(user_and_role(user.login)).to eq(
+          "<img width=\"20\" height=\"20\" alt=\"#{user}\" src=\"/user/icon/#{user.login}?size=20\" " + \
+          "/><a href=\"/user/show/#{user.login}\">#{user.login}</a>"
+        )
+      end
+    end
+
+    context 'with options' do
+      before do
+        User.current = nobody
```
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Why user nobody? I would expect that this belongs to ``for logged in users``

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 3 above)
Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: But does it really make sense to display a role in context of the nobody user? I guess nobody can't have any role anyway ....

Comment (mschnitzer) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: You don't display nobody. You display `user`. Nobody is only the current user. It doesn't matter what you'll set here. If'd set it to a logged in user, the method #user_and_role will just link to profiles then. If he's a guest, it would not link to anything.

Comment (ChrisBr) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: I see!
So the only difference between a logged in user and a nobody user is displaying the link, right?
So I would move everything into a shared_example and then test the shared example for both cases. For the logged in case I would test additional that there is a link.

Also it might sense to use regular expression matcher then.
https://relishapp.com/rspec/rspec-expectations/v/2-2/docs/matchers/match-matcher

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 10 above)
Comment (bgeuken) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: If a user isn't logged in, he is 'nobody'.

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 4 above)
Comment (bgeuken) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: @mschnitzer How is this different to the nobody test?

File: src/api/spec/helpers/webui/webui_helper_spec.rb (hunk 10 above)
Comment (bgeuken) [inline src/api/spec/helpers/webui/webui_helper_spec.rb]: Got it now and I agree with Christian

**Comment by bgeuken** [comment]
Fixed tests an slightly updated them in #3121
